### PR TITLE
TIFF/DNG support

### PR DESCRIPTION
**Pull request by mauricefisher64**
## Changes in this pull request
Support for TIFF & DNG files.  (multi-page and big_tiff not enabled). 

## Checklist
- [x] This PR represents a single feature, fix, or change.
- [x] All applicable changes have been documented.
- [x] Any `TO DO` items (or similar) have been entered as GitHub issues and the link to that issue has been included in a comment.

### CODE REVIEW

**Review by scouten-adobe — APPROVED** [review]
Minor changes to consider, but basically looks good to me.

Caveat: I don't know TIFF format well, so my reading is a bit thin on the specific parsing details.

File: sdk/src/asset_handlers/tiff_io.rs [hunk 1]
```diff
@@ -0,0 +1,1545 @@
...
+const STRIPBYTECOUNTS: u16 = 279;
+const STRIPOFFSETS: u16 = 273;
+const TILEBYTECOUNTS: u16 = 325;
+const TILEOFFSETS: u16 = 324;
+
+const SUBFILES: [u16; 3] = [SUBFILE_TAG, EXIFIFD_TAG, GPSIFD_TAG];
+
+// The type of an IFD entry
+enum IFDEntryType {
+    Byte = 1,       // 8-bit unsigned integer
```
[the remark this answers is not in the document]
Comment (mauricefisher64) [reply sdk/src/asset_handlers/tiff_io.rs:50]: I may look into that but notice there are gaps in the sequence.

File: sdk/src/asset_handlers/tiff_io.rs [hunk 2]
```diff
@@ -0,0 +1,1545 @@
...
+pub struct IfdEntry {
+    entry_tag: u16,
+    entry_type: u16,
+    value_count: u64,
+    value_offset: u64,
+}
+
+// helper enum to know if the IFD requires special handling
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum IFDType {
```
[the remark this answers is not in the document]
Comment (mauricefisher64) [reply sdk/src/asset_handlers/tiff_io.rs:103]: Updated

File: sdk/src/asset_handlers/tiff_io.rs [hunk 3]
```diff
@@ -0,0 +1,1545 @@
...
+    }
+
+    #[allow(dead_code)]
+    pub fn get_tag_mut(&mut self, tag_id: u16) -> Option<&mut IfdEntry> {
+        self.entries.get_mut(&tag_id)
+    }
+}
+
+// Struct to map the contents of a TIFF file
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (mauricefisher64) [reply sdk/src/asset_handlers/tiff_io.rs:133]: Yes, needed until we enable mult-page TIFF support

File: sdk/src/cose_validator.rs [hunk 4]
```diff
@@ -576,6 +566,25 @@ fn get_sign_certs(sign1: &coset::CoseSign1) -> Result<Vec<Vec<u8>>> {
     }
 }
 
+// internal util function to dump the cert chain in PEM format
+#[cfg(not(target_arch = "wasm32"))]
+#[cfg(feature = "file_io")]
+#[allow(dead_code)]
+fn dump_cert_chain(certs: &Vec<Vec<u8>>, output_path: &std::path::Path) -> Result<()> {
```
Comment (gpeacock) [inline sdk/src/cose_validator.rs:573]: could this be
certs: &[Vec<u8>] ?

File: sdk/src/store.rs [hunk 5]
```diff
@@ -3126,4 +3126,79 @@ pub mod tests {
...
+
+        // Move the claim to claims list. Note this is not real, the claims would have to be signed in between commmits
+        store.commit_claim(claim1).unwrap();
+        store.save_to_asset(&ap, &signer, &op).unwrap();
+        store.commit_claim(claim_capture).unwrap();
+        store.save_to_asset(&op, &signer, &op).unwrap();
+        store.commit_claim(claim2).unwrap();
+        store.save_to_asset(&op, &signer, &op).unwrap();
+
+        // write to new file
```
Comment (gpeacock) [inline sdk/src/store.rs:3163]: wrong comment?

File: sdk/src/asset_handlers/tiff_io.rs [hunk 6]
```diff
@@ -0,0 +1,1546 @@
...
+    }
+
+    // move read point to start of entry
+    let decoded_offset = decode_offset(cai_ifd_entry.value_offset, e, big_tiff)?;
+    asset_reader.seek(SeekFrom::Start(decoded_offset))?;
+
+    let manifest_len: usize = usize::value_from(cai_ifd_entry.value_count)
+        .map_err(|_err| Error::InvalidAsset("TIFF/DNG out of range".to_string()))?;
+
+    let mut data = vec![0u8; manifest_len];
```
Comment (gpeacock) [inline sdk/src/asset_handlers/tiff_io.rs:1316]: Nice preallocation work!

File: sdk/src/asset_handlers/tiff_io.rs [hunk 7]
```diff
@@ -0,0 +1,1546 @@
...
+        let entry = IfdClonedEntry {
+            entry_tag: C2PA_TAG,
+            entry_type: C2PA_FIELD_TYPE,
+            value_count: l,
+            value_bytes: store_bytes.to_vec(),
+        };
+
+        tiff_clone_with_tags(&mut temp_file, &mut reader, vec![entry])?;
+
+        std::fs::copy(temp_file.path(), asset_path)?;
```
Comment (gpeacock) [inline sdk/src/asset_handlers/tiff_io.rs:1393]: try a rename first like we do in other places?

File: sdk/src/asset_handlers/tiff_io.rs [hunk 8]
```diff
@@ -0,0 +1,1546 @@
...
+
+        Ok(vec![HashObjectPositions {
+            offset: manifest_offset,
+            length: manifest_len,
+            htype: HashBlockObjectType::Cai,
+        }])
+    }
+
+    fn remove_cai_store(&self, _asset_path: &std::path::Path) -> Result<()> {
+        Ok(())
```
Comment (gpeacock) [inline sdk/src/asset_handlers/tiff_io.rs:1435]: Really seems like it should return an error if it isn't supported

**Review by gpeacock — CHANGES_REQUESTED** [review]
Looks very clean for working with such a nasty format! I added some suggestions.

File: sdk/src/asset_handlers/tiff_io.rs (hunk 8 above)
Comment (mauricefisher64) [reply sdk/src/asset_handlers/tiff_io.rs:1435]: I will add support.  I just overlooked this case.

File: sdk/src/asset_handlers/tiff_io.rs [hunk 9]
```diff
@@ -0,0 +1,1546 @@
...
+        for &mut IfdClonedEntry {
+            value_bytes: ref mut value_bytes_ref,
+            ..
+        } in target_ifd.values_mut()
+        {
+            let data_bytes = if self.big_tiff { 8 } else { 4 };
+
+            if value_bytes_ref.len() > data_bytes {
+                // Start on a WORD boundary
+                //self.pad_word_boundary()?;
```
[the remark this answers is not in the document]
Comment (mauricefisher64) [reply sdk/src/asset_handlers/tiff_io.rs:567]: I found an article say values begin on boundaries but all the examples I see don't seem to require this.  Also I could not find it in the spec.  I will remove to make it more clear.

**Review by gpeacock — APPROVED** [review]

**Pull request merged**